File: Game.py
import pygame
import sys
import random
import math

# Initialize Pygame
pygame.init()

# Constants
SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 0, 0)
PURPLE = (128, 0, 128)
BLUE = (0, 0, 255)
GREEN = (0, 255, 0)
GOLD = (255, 215, 0)
PLAYER_WIDTH = 50
PLAYER_HEIGHT = 50
PLAYER_SPEED = 6
GRAVITY = 0.5
MAX_JUMP = 2
ENEMY_WIDTH = 30
ENEMY_HEIGHT = 30
ENEMY_INITIAL_SPEED = 3
ENEMY_SPEED = 3
BULLET_WIDTH = 8
BULLET_HEIGHT = 8
BULLET_SPEED = 12
MAX_ENEMIES = 5
POWERUP_SIZE = 20
POWERUP_DURATION = 5000  # 5 seconds in milliseconds

# Create the game window
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("Bulletstorm Blitz")

# Player class
class Player(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        self.image = pygame.Surface((PLAYER_WIDTH, PLAYER_HEIGHT))
        self.image.fill(RED)
        self.rect = self.image.get_rect()
        self.rect.center = (SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2)
        self.vel_y = 0
        self.on_ground = False
        self.jumps = 0
        self.score = 0
        self.high_score = 0
        self.level = 1
        self.rapid_fire = False
        self.rapid_fire_timer = 0
        self.shield = False
        self.shield_timer = 0
        self.multiplier = 1
        self.combo_timer = 0

    def update(self, platforms):
        # Apply gravity
        self.vel_y += GRAVITY
        self.rect.y += self.vel_y

        # Check for collisions with platforms
        self.on_ground = False
        for platform in platforms:
            if self.rect.colliderect(platform.rect):
                if self.vel_y > 0:
                    self.on_ground = True
                    self.rect.bottom = platform.rect.top
                    self.vel_y = 0
                    self.jumps = 0
                elif self.vel_y < 0:
                    self.rect.top = platform.rect.bottom
                    self.vel_y = 0

        # Handle player movement
        keys = pygame.key.get_pressed()
        if keys[pygame.K_LEFT] and self.rect.left > 0:
            self.rect.x -= PLAYER_SPEED
        if keys[pygame.K_RIGHT] and self.rect.right < SCREEN_WIDTH:
            self.rect.x += PLAYER_SPEED
        if keys[pygame.K_UP] and (self.on_ground or self.jumps < MAX_JUMP):
            self.vel_y = -12
            self.jumps += 1

        # Update power-up timers
        current_time = pygame.time.get_ticks()
        if self.rapid_fire and current_time > self.rapid_fire_timer:
            self.rapid_fire = False
        if self.shield and current_time > self.shield_timer:
            self.shield = False
        
        # Update combo multiplier
        if current_time > self.combo_timer:
            self.multiplier = 1

    def shoot(self):
        if self.rapid_fire:
            bullet_count = 3
            spread = 15
        else:
            bullet_count = 1
            spread = 0

        bullets = []
        for i in range(bullet_count):
            angle_offset = (i - (bullet_count-1)/2) * spread
            bullet = Bullet(self.rect.centerx, self.rect.centery, angle_offset)
            bullets.append(bullet)
        
        return bullets

# Enemy class with improved AI
class Enemy(pygame.sprite.Sprite):
    def __init__(self, player):
        super().__init__()
        self.image = pygame.Surface((ENEMY_WIDTH, ENEMY_HEIGHT))
        self.image.fill(BLACK)
        self.rect = self.image.get_rect()
        self.spawn_position()
        self.player = player
        self.speed = ENEMY_INITIAL_SPEED
        self.behavior_timer = pygame.time.get_ticks()
        self.behavior = "chase"

    def spawn_position(self):
        side = random.choice(["top", "right", "bottom", "left"])
        if side == "top":
            self.rect.x = random.randint(0, SCREEN_WIDTH - ENEMY_WIDTH)
            self.rect.y = -ENEMY_HEIGHT
        elif side == "right":
            self.rect.x = SCREEN_WIDTH
            self.rect.y = random.randint(0, SCREEN_HEIGHT - ENEMY_HEIGHT)
        elif side == "bottom":
            self.rect.x = random.randint(0, SCREEN_WIDTH - ENEMY_WIDTH)
            self.rect.y = SCREEN_HEIGHT
        else:
            self.rect.x = -ENEMY_WIDTH
            self.rect.y = random.randint(0, SCREEN_HEIGHT - ENEMY_HEIGHT)

    def update(self):
        current_time = pygame.time.get_ticks()
        
        # Change behavior every 3 seconds
        if current_time - self.behavior_timer > 3000:
            self.behavior = random.choice(["chase", "circle", "zigzag"])
            self.behavior_timer = current_time

        if self.behavior == "chase":
            self.chase_player()
        elif self.behavior == "circle":
            self.circle_player()
        elif self.behavior == "zigzag":
            self.zigzag_movement()

    def chase_player(self):
        dx = self.player.rect.centerx - self.rect.centerx
        dy = self.player.rect.centery - self.rect.centery
        dist = math.sqrt(dx * dx + dy * dy)
        if dist != 0:
            self.rect.x += (dx / dist) * self.speed
            self.rect.y += (dy / dist) * self.speed

    def circle_player(self):
        angle = pygame.time.get_ticks() / 500  # Rotation speed
        radius = 100  # Circle radius
        self.rect.x = self.player.rect.centerx + math.cos(angle) * radius - ENEMY_WIDTH/2
        self.rect.y = self.player.rect.centery + math.sin(angle) * radius - ENEMY_HEIGHT/2

    def zigzag_movement(self):
        self.chase_player()
        self.rect.x += math.sin(pygame.time.get_ticks() / 200) * 5

# Bullet class
class Bullet(pygame.sprite.Sprite):
    def __init__(self, x, y, angle_offset=0):
        super().__init__()
        self.image = pygame.Surface((BULLET_WIDTH, BULLET_HEIGHT))
        self.image.fill(BLUE)
        self.rect = self.image.get_rect()
        self.rect.center = (x, y)
        
        # Find nearest enemy
        nearest_enemy = None
        min_distance = float('inf')
        for enemy in enemies_group:
            dx = enemy.rect.centerx - x
            dy = enemy.rect.centery - y
            distance = math.sqrt(dx ** 2 + dy ** 2)
            if distance < min_distance:
                min_distance = distance
                nearest_enemy = enemy

        if nearest_enemy:
            dx = nearest_enemy.rect.centerx - x
            dy = nearest_enemy.rect.centery - y
            angle = math.atan2(dy, dx)
            angle += math.radians(angle_offset)  # Add spread angle
            self.velocity_x = math.cos(angle) * BULLET_SPEED
            self.velocity_y = math.sin(angle) * BULLET_SPEED
        else:
            self.velocity_x = 0
            self.velocity_y = -BULLET_SPEED

    def update(self):
        self.rect.x += self.velocity_x
        self.rect.y += self.velocity_y
        
        # Remove if off screen
        if (self.rect.right < 0 or self.rect.left > SCREEN_WIDTH or
            self.rect.bottom < 0 or self.rect.top > SCREEN_HEIGHT):
            self.kill()

# Platform class
class Platform(pygame.sprite.Sprite):
    def __init__(self, x, y, width, height):
        super().__init__()
        self.image = pygame.Surface((width, height))
        self.image.fill(BLACK)
        self.rect = self.image.get_rect(topleft=(x, y))

# PowerUp class
class PowerUp(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        self.type = random.choice(["rapid_fire", "shield", "multiplier"])
        self.image = pygame.Surface((POWERUP_SIZE, POWERUP_SIZE))
        if self.type == "rapid_fire":
            self.image.fill(GOLD)
        elif self.type == "shield":
            self.image.fill(BLUE)
        else:
            self.image.fill(GREEN)
        self.rect = self.image.get_rect()
        self.rect.x = random.randint(0, SCREEN_WIDTH - POWERUP_SIZE)
        self.rect.y = random.randint(0, SCREEN_HEIGHT - POWERUP_SIZE)

# Create sprite groups
player = Player()
enemies_group = pygame.sprite.Group()
bullets_group = pygame.sprite.Group()
platforms_group = pygame.sprite.Group()
powerups_group = pygame.sprite.Group()

# Create platforms
platforms = [
    Platform(0, SCREEN_HEIGHT - 50, SCREEN_WIDTH, 50),  # Ground
    Platform(SCREEN_WIDTH // 4, SCREEN_HEIGHT // 2, 200, 20),  # Left platform
    Platform(SCREEN_WIDTH * 3 // 4 - 200, SCREEN_HEIGHT // 2, 200, 20),  # Right platform
    Platform(SCREEN_WIDTH // 2 - 100, SCREEN_HEIGHT * 3 // 4, 200, 20),  # Bottom middle platform
]
platforms_group.add(platforms)

def spawn_powerup():
    if random.random() < 0.02 and len(powerups_group) < 3:  # 2% chance per frame, max 3 powerups
        powerups_group.add(PowerUp())

def new_level():
    bullets_group.empty()
    powerups_group.empty()
    player.level += 1
    for _ in range(min(MAX_ENEMIES + player.level - 1, 10)):  # Increase enemies with level, max 10
        enemy = Enemy(player)
        enemy.speed = min(ENEMY_SPEED + (player.level - 1) * 0.5, 7)  # Increase speed with level, max 7
        enemies_group.add(enemy)
    player.rect.center = (SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2)

# Main game loop
clock = pygame.time.Clock()
last_shot_time = 0
shot_delay = 250  # Milliseconds between shots
running = True

while running:
    current_time = pygame.time.get_ticks()
    
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                if current_time - last_shot_time > (shot_delay / 2 if player.rapid_fire else shot_delay):
                    bullets = player.shoot()
                    bullets_group.add(bullets)
                    last_shot_time = current_time

    # Update
    player.update(platforms_group)
    enemies_group.update()
    bullets_group.update()

    # Spawn power-ups
    spawn_powerup()

    # Check for collisions between bullets and enemies
    for bullet in bullets_group:
        enemy_hit = pygame.sprite.spritecollideany(bullet, enemies_group)
        if enemy_hit:
            bullet.kill()
            enemy_hit.kill()
            player.score += 100 * player.multiplier
            player.combo_timer = current_time + 2000  # 2 second combo window
            player.multiplier = min(player.multiplier + 0.5, 4)  # Max 4x multiplier

    # Check for collisions between player and power-ups
    powerup_hit = pygame.sprite.spritecollideany(player, powerups_group)
    if powerup_hit:
        if powerup_hit.type == "rapid_fire":
            player.rapid_fire = True
            player.rapid_fire_timer = current_time + POWERUP_DURATION
        elif powerup_hit.type == "shield":
            player.shield = True
            player.shield_timer = current_time + POWERUP_DURATION
        else:  # multiplier
            player.multiplier *= 2
        powerup_hit.kill()

    # Check for collisions between player and enemies
    if not player.shield and pygame.sprite.spritecollideany(player, enemies_group):
        player.high_score = max(player.high_score, player.score)
        player.score = 0
        player.level = 1
        player.multiplier = 1
        enemies_group.empty()
        new_level()

    # Check if all enemies are defeated
    if len(enemies_group) == 0:
        new_level()

    # Draw
    screen.fill(WHITE)
    
    # Draw platforms
    platforms_group.draw(screen)
    
    # Draw power-ups
    powerups_group.draw(screen)
    
    # Draw player with shield effect
    if player.shield:
        pygame.draw.circle(screen, BLUE, player.rect.center, max(PLAYER_WIDTH, PLAYER_HEIGHT) // 2 + 5, 2)
    screen.blit(player.image, player.rect)
    
    # Draw enemies and bullets
    enemies_group.draw(screen)
    bullets_group.draw(screen)

    # Display score, level, and high score
    font = pygame.font.Font(None, 36)
    score_text = font.render(f"Score: {player.score}", True, BLACK)
    level_text = font.render(f"Level: {player.level}", True, BLACK)
    high_score_text = font.render(f"High Score: {player.high_score}", True, BLACK)
    multiplier_text = font.render(f"Multiplier: {player.multiplier:.1f}x", True, BLACK)
    
    screen.blit(score_text, (10, 10))
    screen.blit(level_text, (10, 50))
    screen.blit(high_score_text, (10, 90))
    screen.blit(multiplier_text, (10, 130))

    # Display active power-ups
    if player.rapid_fire:
        rapid_fire_text = font.render("Rapid Fire!", True, GOLD)
        screen.blit(rapid_fire_text, (SCREEN_WIDTH - 150, 10))
    if player.shield:
        shield_text = font.render("Shield!", True, BLUE)
        screen.blit(shield_text, (SCREEN_WIDTH - 150, 50))

    pygame.display.flip()
    clock.tick(60)

pygame.quit()
sys.exit()

File: game.js
// Constants
const SCREEN_WIDTH = 800;
const SCREEN_HEIGHT = 600;
const PLAYER_WIDTH = 50;
const PLAYER_HEIGHT = 50;
const PLAYER_SPEED = 6;
const GRAVITY = 0.5;
const MAX_JUMP = 2;
const ENEMY_WIDTH = 30;
const ENEMY_HEIGHT = 30;
const ENEMY_SPEED = 3;
const BULLET_WIDTH = 8;
const BULLET_HEIGHT = 8;
const BULLET_SPEED = 12;
const MAX_ENEMIES = 5;
const POWERUP_SIZE = 20;
const POWERUP_DURATION = 5000;

// Colors
const COLORS = {
    WHITE: '#FFFFFF',
    BLACK: '#000000',
    RED: '#FF0000',
    BLUE: '#0000FF',
    GREEN: '#00FF00',
    GOLD: '#FFD700',
    PURPLE: '#800080'
};

// Game state
let player;
let enemies = [];
let bullets = [];
let platforms = [];
let powerups = [];
let lastTime = 0;
let lastShotTime = 0;

// Canvas setup
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');

class Player {
    constructor() {
        this.width = PLAYER_WIDTH;
        this.height = PLAYER_HEIGHT;
        this.x = SCREEN_WIDTH / 2 - this.width / 2;
        this.y = SCREEN_HEIGHT / 2 - this.height / 2;
        this.velY = 0;
        this.onGround = false;
        this.jumps = 0;
        this.score = 0;
        this.highScore = 0;
        this.level = 1;
        this.rapidFire = false;
        this.rapidFireTimer = 0;
        this.shield = false;
        this.shieldTimer = 0;
        this.multiplier = 1;
        this.comboTimer = 0;
    }

    update() {
        // Apply gravity
        this.velY += GRAVITY;
        this.y += this.velY;

        // Platform collision
        this.onGround = false;
        for (const platform of platforms) {
            if (this.collidesWith(platform)) {
                if (this.velY > 0) {
                    this.onGround = true;
                    this.y = platform.y - this.height;
                    this.velY = 0;
                    this.jumps = 0;
                } else if (this.velY < 0) {
                    this.y = platform.y + platform.height;
                    this.velY = 0;
                }
            }
        }

        // Movement
        if (keys.ArrowLeft && this.x > 0) {
            this.x -= PLAYER_SPEED;
        }
        if (keys.ArrowRight && this.x < SCREEN_WIDTH - this.width) {
            this.x += PLAYER_SPEED;
        }
        if (keys.ArrowUp && (this.onGround || this.jumps < MAX_JUMP)) {
            this.velY = -12;
            this.jumps++;
        }

        // Update power-up timers
        const currentTime = Date.now();
        if (this.rapidFire && currentTime > this.rapidFireTimer) {
            this.rapidFire = false;
        }
        if (this.shield && currentTime > this.shieldTimer) {
            this.shield = false;
        }
        if (currentTime > this.comboTimer) {
            this.multiplier = 1;
        }
    }

    shoot() {
        const currentTime = Date.now();
        if (currentTime - lastShotTime < (this.rapidFire ? 125 : 250)) return;

        const bulletCount = this.rapidFire ? 3 : 1;
        const spread = this.rapidFire ? 15 : 0;

        for (let i = 0; i < bulletCount; i++) {
            const angleOffset = (i - (bulletCount - 1) / 2) * spread;
            bullets.push(new Bullet(
                this.x + this.width / 2,
                this.y + this.height / 2,
                angleOffset
            ));
        }

        lastShotTime = currentTime;
    }

    collidesWith(obj) {
        return this.x < obj.x + obj.width &&
               this.x + this.width > obj.x &&
               this.y < obj.y + obj.height &&
               this.y + this.height > obj.y;
    }

    draw() {
        ctx.fillStyle = COLORS.RED;
        ctx.fillRect(this.x, this.y, this.width, this.height);
        
        if (this.shield) {
            ctx.strokeStyle = COLORS.BLUE;
            ctx.beginPath();
            ctx.arc(
                this.x + this.width / 2,
                this.y + this.height / 2,
                Math.max(this.width, this.height) / 2 + 5,
                0,
                Math.PI * 2
            );
            ctx.stroke();
        }
    }
}

class Enemy {
    constructor() {
        this.width = ENEMY_WIDTH;
        this.height = ENEMY_HEIGHT;
        this.speed = ENEMY_SPEED;
        this.spawnPosition();
        this.behaviorTimer = Date.now();
        this.behavior = "chase";
    }

    spawnPosition() {
        const side = Math.floor(Math.random() * 4);
        switch(side) {
            case 0: // top
                this.x = Math.random() * (SCREEN_WIDTH - this.width);
                this.y = -this.height;
                break;
            case 1: // right
                this.x = SCREEN_WIDTH;
                this.y = Math.random() * (SCREEN_HEIGHT - this.height);
                break;
            case 2: // bottom
                this.x = Math.random() * (SCREEN_WIDTH - this.width);
                this.y = SCREEN_HEIGHT;
                break;
            case 3: // left
                this.x = -this.width;
                this.y = Math.random() * (SCREEN_HEIGHT - this.height);
                break;
        }
    }

    update() {
        const currentTime = Date.now();
        
        if (currentTime - this.behaviorTimer > 3000) {
            this.behavior = ["chase", "circle", "zigzag"][Math.floor(Math.random() * 3)];
            this.behaviorTimer = currentTime;
        }

        switch(this.behavior) {
            case "chase":
                this.chasePlayer();
                break;
            case "circle":
                this.circlePlayer();
                break;
            case "zigzag":
                this.zigzagMovement();
                break;
        }
    }

    chasePlayer() {
        const dx = player.x - this.x;
        const dy = player.y - this.y;
        const dist = Math.sqrt(dx * dx + dy * dy);
        if (dist !== 0) {
            this.x += (dx / dist) * this.speed;
            this.y += (dy / dist) * this.speed;
        }
    }

    circlePlayer() {
        const angle = Date.now() / 500;
        const radius = 100;
        this.x = player.x + Math.cos(angle) * radius - this.width / 2;
        this.y = player.y + Math.sin(angle) * radius - this.height / 2;
    }

    zigzagMovement() {
        this.chasePlayer();
        this.x += Math.sin(Date.now() / 200) * 5;
    }

    draw() {
        ctx.fillStyle = COLORS.BLACK;
        ctx.fillRect(this.x, this.y, this.width, this.height);
    }
}

class Bullet {
    constructor(x, y, angleOffset = 0) {
        this.width = BULLET_WIDTH;
        this.height = BULLET_HEIGHT;
        this.x = x;
        this.y = y;

        let nearestEnemy = null;
        let minDistance = Infinity;
        
        for (const enemy of enemies) {
            const dx = enemy.x - x;
            const dy = enemy.y - y;
            const distance = Math.sqrt(dx * dx + dy * dy);
            if (distance < minDistance) {
                minDistance = distance;
                nearestEnemy = enemy;
            }
        }

        if (nearestEnemy) {
            const dx = nearestEnemy.x - x;
            const dy = nearestEnemy.y - y;
            const angle = Math.atan2(dy, dx) + (angleOffset * Math.PI / 180);
            this.velocityX = Math.cos(angle) * BULLET_SPEED;
            this.velocityY = Math.sin(angle) * BULLET_SPEED;
        } else {
            this.velocityX = 0;
            this.velocityY = -BULLET_SPEED;
        }
    }

    update() {
        this.x += this.velocityX;
        this.y += this.velocityY;
    }

    isOffScreen() {
        return this.x < 0 || this.x > SCREEN_WIDTH ||
               this.y < 0 || this.y > SCREEN_HEIGHT;
    }

    draw() {
        ctx.fillStyle = COLORS.BLUE;
        ctx.fillRect(this.x, this.y, this.width, this.height);
    }
}

class PowerUp {
    constructor() {
        this.size = POWERUP_SIZE;
        this.type = ["rapid_fire", "shield", "multiplier"][Math.floor(Math.random() * 3)];
        this.x = Math.random() * (SCREEN_WIDTH - this.size);
        this.y = Math.random() * (SCREEN_HEIGHT - this.size);
    }

    draw() {
        ctx.fillStyle = this.type === "rapid_fire" ? COLORS.GOLD :
                       this.type === "shield" ? COLORS.BLUE :
                       COLORS.GREEN;
        ctx.fillRect(this.x, this.y, this.size, this.size);
    }
}

// Input handling
const keys = {};
document.addEventListener('keydown', (e) => {
    keys[e.code] = true;
    if (e.code === 'Space') {
        player.shoot();
    }
});
document.addEventListener('keyup', (e) => {
    keys[e.code] = false;
});

function spawnPowerup() {
    if (Math.random() < 0.02 && powerups.length < 3) {
        powerups.push(new PowerUp());
    }
}

function newLevel() {
    bullets = [];
    powerups = [];
    player.level++;
    enemies = [];
    const enemyCount = Math.min(MAX_ENEMIES + player.level - 1, 10);
    for (let i = 0; i < enemyCount; i++) {
        const enemy = new Enemy();
        enemy.speed = Math.min(ENEMY_SPEED + (player.level - 1) * 0.5, 7);
        enemies.push(enemy);
    }
    player.x = SCREEN_WIDTH / 2 - player.width / 2;
    player.y = SCREEN_HEIGHT / 2 - player.height / 2;
}

function init() {
    player = new Player();
    
    // Create platforms
    platforms = [
        { x: 0, y: SCREEN_HEIGHT - 50, width: SCREEN_WIDTH, height: 50 },
        { x: SCREEN_WIDTH / 4, y: SCREEN_HEIGHT / 2, width: 200, height: 20 },
        { x: SCREEN_WIDTH * 3/4 - 200, y: SCREEN_HEIGHT / 2, width: 200, height: 20 },
        { x: SCREEN_WIDTH / 2 - 100, y: SCREEN_HEIGHT * 3/4, width: 200, height: 20 }
    ];
    
    newLevel();
}

function update() {
    player.update();
    
    // Update enemies
    enemies.forEach(enemy => enemy.update());
    
    // Update bullets
    bullets = bullets.filter(bullet => {
        bullet.update();
        return !bullet.isOffScreen();
    });
    
    // Spawn power-ups
    spawnPowerup();
    
    // Check bullet-enemy collisions
    bullets = bullets.filter(bullet => {
        let hit = false;
        enemies = enemies.filter(enemy => {
            if (!hit && bullet.x < enemy.x + enemy.width &&
                bullet.x + bullet.width > enemy.x &&
                bullet.y < enemy.y + enemy.height &&
                bullet.y + bullet.height > enemy.y) {
                hit = true;
                player.score += 100 * player.multiplier;
                player.comboTimer = Date.now() + 2000;
                player.multiplier = Math.min(player.multiplier + 0.5, 4);
                return false;
            }
            return true;
        });
        return !hit;
    });
    
    // Check player-powerup collisions
    powerups = powerups.filter(powerup => {
        if (player.x < powerup.x + powerup.size &&
            player.x + player.width > powerup.x &&
            player.y < powerup.y + powerup.size &&
            player.y + player.height > powerup.y) {
            const currentTime = Date.now();
            switch(powerup.type) {
                case "rapid_fire":
                    player.rapidFire = true;
                    player.rapidFireTimer = currentTime + POWERUP_DURATION;
                    break;
                case "shield":
                    player.shield = true;
                    player.shieldTimer = currentTime + POWERUP_DURATION;
                    break;
                case "multiplier":
                    player.multiplier *= 2;
                    break;
            }
            return false;
        }
        return true;
    });
    
    // Check player-enemy collisions
    if (!player.shield && enemies.some(enemy => player.collidesWith(enemy))) {
        player.highScore = Math.max(player.highScore, player.score);
        player.score = 0;
        player.level = 1;
        player.multiplier = 1;
        enemies = [];
        newLevel();
    }
    
    // Check if level completed
    if (enemies.length === 0) {
        newLevel();
    }
}

function draw() {
    // Clear canvas
    ctx.fillStyle = COLORS.WHITE;
    ctx.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
    
    // Draw platforms
    ctx.fillStyle = COLORS.BLACK;
    platforms.forEach(platform => {
        ctx.fillRect(platform.x, platform.y, platform.width, platform.height);
    });
    
    // Draw power-ups
    powerups.forEach(powerup => powerup.draw());
    
    // Draw player
    player.draw();
    
    // Draw enemies
    enemies.forEach(enemy => enemy.draw());
    
    // Draw bullets
    bullets.forEach(bullet => bullet.draw());
    
    // Draw UI
    ctx.fillStyle = COLORS.BLACK;
    ctx.font = '36px Arial';
    ctx.fillText(`Score: ${player.score}`, 10, 40);
    ctx.fillText(`Level: ${player.level}`, 10, 80);
    ctx.fillText(`High Score: ${player.highScore}`, 10, 120);
    ctx.fillText(`Multiplier: ${player.multiplier.toFixed(1)}x`, 10, 160);
    
    if (player.rapidFire) {
        ctx.fillStyle = COLORS.GOLD;
        ctx.fillText('Rapid Fire!', SCREEN_WIDTH - 150, 40);
    }
    if (player.shield) {
        ctx.fillStyle = COLORS.BLUE;
        ctx.fillText('Shield!', SCREEN_WIDTH - 150, 80);
    }
}

function gameLoop(timestamp) {
    const deltaTime = timestamp - lastTime;
    lastTime = timestamp;
    
    update();
    draw();
    
    requestAnimationFrame(gameLoop);
}

// Start the game
init();
requestAnimationFrame(gameLoop);